pkg/operator/operator/autoscaler.go

```diff
@@ -0,0 +1,263 @@
+			if request < currentReplicas {
+				request = currentReplicas
+			}
+		} else if downscaleStabilizationFloor != nil && request < *downscaleStabilizationFloor {
+			request = *downscaleStabilizationFloor
+		}
+
+		upscaleStabilizationCeil := recs.minSince(autoscalingSpec.UpscaleStabilizationPeriod)
+		if time.Since(startTime) < autoscalingSpec.UpscaleStabilizationPeriod {
+			if request > currentReplicas {
+				request = currentReplicas
+			}
+		} else if upscaleStabilizationCeil != nil && request > *upscaleStabilizationCeil {
+			request = *upscaleStabilizationCeil
+		}
+
+		log.Printf("%s autoscaler tick: total_in_flight=%s, raw_recommendation=%s, downscale_factor_floor=%d, upscale_factor_ceil=%d, min_replicas=%d, max_replicas=%d, recommendation=%d, downscale_stabilization_floor=%s, upscale_stabilization_ceil=%s, current_replicas=%d, request=%d", apiName, s.Round(*totalInFlight, 2, 0), s.Round(rawRecommendation, 2, 0), downscaleFactorFloor, upscaleFactorCeil, autoscalingSpec.MinReplicas, autoscalingSpec.MaxReplicas, recommendation, s.ObjFlatNoQuotes(downscaleStabilizationFloor), s.ObjFlatNoQuotes(upscaleStabilizationCeil), currentReplicas, request)
+
+		if currentReplicas != request {
+			log.Printf("%s autoscaling event: %d -> %d", apiName, currentReplicas, request)
+
+			deployment, err := config.K8s.GetDeployment(initialDeployment.Name)
+			if err != nil {
+				return err
+			}
+
+			deployment.Spec.Replicas = &request
+
+			if _, err := config.K8s.UpdateDeployment(deployment); err != nil {
+				return err
+			}
+
+			currentReplicas = request
+		}
+
+		return nil
+	}, nil
+}
+
+func getInflightRequests(apiName string, window time.Duration, tickInterval time.Duration) (*float64, error) {
+	endTime := time.Now().Truncate(time.Second)
+	startTime := endTime.Add(-2 * window)
+	metricsDataQuery := cloudwatch.GetMetricDataInput{
+		EndTime:   &endTime,
+		StartTime: &startTime,
+		MetricDataQueries: []*cloudwatch.MetricDataQuery{
+			{
+				Id:    aws.String("inflight"),
+				Label: aws.String("InFlight"),
+				MetricStat: &cloudwatch.MetricStat{
+					Metric: &cloudwatch.Metric{
+						Namespace:  aws.String(config.Cluster.LogGroup),
+						MetricName: aws.String("in-flight"),
+						Dimensions: []*cloudwatch.Dimension{
+							{
+								Name:  aws.String("apiName"),
+								Value: aws.String(apiName),
+							},
+						},
+					},
+					Stat:   aws.String("Sum"),
+					Period: aws.Int64(10),
+				},
+			},
+		},
+	}
+
+	output, err := config.AWS.CloudWatchMetrics().GetMetricData(&metricsDataQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(output.MetricDataResults[0].Timestamps) == 0 {
+		return nil, nil
+	}
+
+	timestampCounter := -1
+	for i, timeStamp := range output.MetricDataResults[0].Timestamps {
+		if endTime.Sub(*timeStamp) < 20*time.Second {
+			timestampCounter = i
+		} else {
+			break
+		}
+	}
+
+	if timestampCounter == -1 {
+		return nil, nil // no metrics were available in the last 2 tick intervals
```

Comment:
maybe return err here indicating metrics were not found?

Reply:
same comment as above